src/Compiler/Driver/CompilerConfig.fs

```diff
@@ -91,24 +92,41 @@ let ResolveFileUsingPaths (paths, m, fileName) =
         let searchMessage = String.concat "\n " paths
         raise (FileNameNotResolved(fileName, searchMessage, m))
 
-let GetWarningNumber (m, warningNumber: string, prefixSupported) =
-    try
-        let warningNumber =
-            if warningNumber.StartsWithOrdinal "FS" then
-                if prefixSupported then
-                    warningNumber.Substring 2
-                else
-                    raise (new ArgumentException())
-            else
-                warningNumber
+let private rxOptions = RegexOptions.Compiled ||| RegexOptions.CultureInvariant
```

Comment:
Here is an idea:

Instead of many individually regexes, where each is eagerly compiled (which happens at runtime via IL emit = CPU work), create one or two unified regex with capture groups.

And then do the branching, either via pattern match or via a fallback chain of if-else as it is done now, for the results of those named captures.

This will save time for:
- All invocations of fsc independent of code ever using these constructs, since regex creation is happening eagerly
- Processing of the hash directives, since there is only one regex processing instead of possibly many in the chain of if-else's

Reply:
Thanks for bringing this up. I used the "compiled" option too carelessly.
I did some benchmarking now. For regular expressions and matches of the given complexity, I have roughly
- 0.5 microseconds per match
- 3.5 microseconds per regex creation (uncompiled)
- 3 milliseconds per compiled regex

Given that even in this big repository we have only some 700 #nowarn directives, compilation does not make sense. And module startup code should be avoided anyway as much as possible.

So, for now, I just went for creating the regex for every match. I believe this makes sense, given the numbers above. (Note also that for every warning number only one match is needed, except for the "NU5125" case (2 matches).) Using named capture groups is a nice idea, but I think the resulting regex would become too complex. The single ones are already not that easy to scan. 
I can give it a try though if you want.